### PR TITLE
Apply validator

### PR DESCRIPTION
**Pull request by Ruins7**
implementation of apply validator & show current validator list

### CODE REVIEW

File: app/controller.go [hunk 1]
```diff
@@ -165,9 +170,16 @@ func (app *App) commitor() commitor {
 	return func() ResponseCommit {
 
 		// Commit any pending changes.
-		hash, ver := app.Context.balances.Commit()
+		hashb, verb := app.Context.balances.Commit()
```
Comment (reblws) [inline app/controller.go]: what's verb and verv?

File: app/controller.go [hunk 2]
```diff
@@ -165,9 +170,16 @@ func (app *App) commitor() commitor {
...
+
+		hashv, verv := app.Context.validators.Commit()
+
+		apphash := &appHash{}
+		apphash.Hashes = append(apphash.Hashes, hashb, hashv)
 
-		app.logger.Debugf("Committed New Block hash[%s], version[%d]", hex.EncodeToString(hash), ver)
+		_, _ = verb, verv
```
Comment (reblws) [inline app/controller.go]: ?

File: app/controller.go (hunk 1 above)
Comment (alexjlan) [reply app/controller.go]: version for balance and version for validators

File: app/controller.go [hunk 3]
```diff
@@ -177,3 +189,24 @@ func (app *App) commitor() commitor {
...
+	}
+	return
+}
+
+//todo: make appHash to use a commiter function to finish the commit and hashing for all the store that passed
+type appHash struct {
+	Hashes [][]byte `json:"hashes"`
+}
+
+func (ah *appHash) hash() []byte {
```
Comment (reblws) [inline app/controller.go]: Maybe you could add a generic hasher to `utils/` ?

File: data/accounts/wallet.go [hunk 4]
```diff
@@ -71,13 +72,13 @@ func (ws *WalletStore) Add(account Account) error {
 
 	exist := session.Exists(account.Address().Bytes())
 	if exist {
-		return errors.New("account already exist: " + string(account.Address()))
+		return errors.New("account already exist: " + hex.EncodeToString(account.Address()))
```
Comment (reblws) [inline data/accounts/wallet.go]: Use `account.Address().String()`

File: data/accounts/wallet.go [hunk 5]
```diff
@@ -94,7 +95,7 @@ func (ws *WalletStore) Delete(account Account) error {
 
 	exist := session.Exists(account.Address().Bytes())
 	if !exist {
-		return errors.New("account does not exist: " + string(account.Address()))
+		return errors.New("account already exist: " + hex.EncodeToString(account.Address()))
```
Comment (reblws) [inline data/accounts/wallet.go]: .String()

File: identity/validator_queue_test.go [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+
+func TestMain(m *testing.M) {
+	setup()
+	code := m.Run()
+	shutdown()
+	os.Exit(code)
+}
+
+func subtestsetup(funName string) {
+	logger.Debug("This is subtest setting up for", funName)
```
Comment (reblws) [inline identity/validator_queue_test.go]: Nothing gets logged to test, when running `go test` the output will be suppressed

File: identity/validator_queue_test.go [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+)
+
+var vq1 = utils.NewQueued([]byte("a"), 200, 1)
+var vq2 = utils.NewQueued([]byte("b"), 1000, 2)
+var vq3 = utils.NewQueued([]byte("c"), 300, 3)
+var vq4 = utils.NewQueued([]byte("d"), 2000, 4)
+
+var Vq *ValidatorQueue
+
+func setup() {
```
Comment (reblws) [inline identity/validator_queue_test.go]: Only use setup and teardown for side-effects (e.g. file I/O, database setup, etc.). Logs don't do anything.

File: identity/validator_set.go [hunk 8]
```diff
@@ -0,0 +1,243 @@
...
+		err = vs.ChainState.Set(h.Address().Bytes(), validator.Bytes())
+		if err != nil {
+			return req.Validators, errors.New("failed to add initial validators")
+		}
+		validatorUpdates = append(validatorUpdates, v)
+	}
+	return validatorUpdates, nil
+}
+
+//setup the validators according to begin block
```
Comment (reblws) [inline identity/validator_set.go]: Space before comments

File: identity/validator_set.go [hunk 9]
```diff
@@ -0,0 +1,243 @@
...
+
+	return err
+}
+
+//get validators set
+func (vs *ValidatorStore) GetValidatorSet() ([]Validator, error) {
+
+	validatorSet := make([]Validator, 0)
+	vs.ChainState.Iterate(func(key, value []byte) bool {
+		validator := (&Validator{}).FromBytes(value)
```
Comment (reblws) [inline identity/validator_set.go]: `FromBytes` implementation should expose an error.

File: identity/validator_test.go [hunk 10]
```diff
@@ -0,0 +1,32 @@
...
+	"math/big"
+	"testing"
+
+	"github.com/Oneledger/protocol/data/balance"
+	"github.com/Oneledger/protocol/data/keys"
+	"github.com/stretchr/testify/assert"
+)
+
+var v = &Validator{
+	Address:      []byte("89507C7ABC6D1E9124FE94101A0AB38D5085E15A"),
```
Comment (reblws) [inline identity/validator_test.go]: You need to decode this hex properly using `hex.Decode
String`

File: identity/validator_test.go [hunk 11]
```diff
@@ -0,0 +1,32 @@
...
+
+func TestValidator_Bytes(t *testing.T) {
+	t.Run("run Bytes test case", func(t *testing.T) {
+		assert.NotEqual(t, []byte{}, v.Bytes())
+	})
+
+}
+
+func TestValidator_FromBytes(t *testing.T) {
+	t.Run("run FromBytes test case", func(t *testing.T) {
```
Comment (reblws) [inline identity/validator_test.go]: You don't need to run subtests here since it's just a single test

File: action/apply_validator.go [hunk 12]
```diff
@@ -0,0 +1,162 @@
...
+
+	if coin.Currency.Name != "VT" {
+		return false, ErrInvalidAmount
+	}
+
+	return true, nil
+}
+
+func (a applyTx) ProcessCheck(ctx *Context, msg Msg, fee Fee) (bool, Response) {
+	apply := msg.(*ApplyValidator)
```
Comment (reblws) [inline action/apply_validator.go]: Check for ok: `apply, ok := msg.(*ApplyValidator)`

File: app/controller.go (hunk 2 above)
Comment (Ruins7) [reply app/controller.go]: @alexjlan

**Pull request merged**